# Hotel Review Sentiment Analysis: 
- Performed the sentoment analysis using vader setiment scoring technique

### VaderSentiment technique:

In [ ]:
import pandas as pd
import nltk
from gensim.models import Word2Vec
from nltk.corpus import stopwords

In [69]:
data = pd.read_csv('train.csv')
df = data[:1000]

In [49]:
df.shape

(200, 5)

In [4]:
from keras.preprocessing.text import text_to_word_sequence
#Applying word level tokenization and stroing into a column named, 'word_tokens'
word_tokens = []
for i in df['Description']:
  word_tokens.append(text_to_word_sequence(i))
df['word_tokens'] = word_tokens
df.head()

User_ID                                        Description  \
0  id10326  The room was kind of clean but had a VERY stro...   
1  id10327  I stayed at the Crown Plaza April -- - April -...   
2  id10328  I booked this hotel through Hotwire at the low...   
3  id10329  Stayed here with husband and sons on the way t...   
4  id10330  My girlfriends and I stayed here to celebrate ...   

        Browser_Used Device_Used Is_Response  \
0               Edge      Mobile   not happy   
1  Internet Explorer      Mobile   not happy   
2            Mozilla      Tablet   not happy   
3   InternetExplorer     Desktop       happy   
4               Edge      Tablet   not happy   

                                         word_tokens  
0  [the, room, was, kind, of, clean, but, had, a,...  
1  [i, stayed, at, the, crown, plaza, april, apri...  
2  [i, booked, this, hotel, through, hotwire, at,...  
3  [stayed, here, with, husband, and, sons, on, t...  
4  [my, girlfriends, and, i, stayed, here, to, ce...

In [7]:
#Stop words combining step:

#get all the stop words from nltk library:
import nltk
nltk.download('stopwords')
stop = nltk.corpus.stopwords.words('english')

#get all the stop words from sklearn library:
from sklearn.feature_extraction.text import ENGLISH_STOP_WORDS as sklearn_stop

#now combine the stop words
combined_stop = list(set(stop).union(sklearn_stop))


#Stop words removal step: After removing stop wird storing the data into a list named, 'clean data'
clean_data = []

for sen in df['word_tokens']:
    for words in sen:
        if words in combined_stop:
            sen.remove(words)
    clean_data.append(sen)

#print only the 1st record after removing the stop words
clean_data[0]   

[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\91842\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


['room',
 'kind',
 'clean',
 'had',
 'very',
 'strong',
 'smell',
 'dogs',
 'generally',
 'average',
 'ok',
 'a',
 'overnight',
 'stay',
 "you're",
 'too',
 'fussy',
 'consider',
 'staying',
 'if',
 'price',
 'right',
 'breakfast',
 'free',
 'just',
 'better',
 'nothing']

In [9]:
#Creting the column named, 'vader_sentiment' for each review 
score = []
from vaderSentiment.vaderSentiment import SentimentIntensityAnalyzer
sa = SentimentIntensityAnalyzer()
for doc in clean_data:
    for review in doc:
        scores = sa.polarity_scores(review)
    score.append(scores)
df['vader_sentiment'] = score


#Storing only the compound score to a column named, 'compound_scores'
compound_scores = []

for i in df.vader_sentiment:
    for key,value in i.items():
        if key=='compound':
            compound_scores.append(value)
df['compound_scores'] = compound_scores


#printing the firts 30 reviews with their vader sentiment score details
df[:30]

User_ID                                        Description  \
0   id10326  The room was kind of clean but had a VERY stro...   
1   id10327  I stayed at the Crown Plaza April -- - April -...   
2   id10328  I booked this hotel through Hotwire at the low...   
3   id10329  Stayed here with husband and sons on the way t...   
4   id10330  My girlfriends and I stayed here to celebrate ...   
5   id10331  We had - rooms. One was very nice and clearly ...   
6   id10332  My husband and I have stayed in this hotel a f...   
7   id10333  My wife & I stayed in this glorious city a whi...   
8   id10334  My boyfriend and I stayed at the Fairmont on a...   
9   id10335  Wonderful staff, great location, but it was de...   
10  id10336  Steps off Times Square, nice rooms, stayed - n...   
11  id10337  Me, the Wife and - kids stayed here on Valenti...   
12  id10338  We stay at the Jolly Madison over the Xmas per...   
13  id10339  I highly recommend the Hawthorne Terrace as an...   
14  id10340  I found the hotel clean and nicely located. Go...   
15  id10341  Stayed at the Elan from --th to --th October a...   
16  id10342  Priceline sent us to this hotel after acceptin...   
17  id10343  Old and cheap furnitures,our chair was simply ...   
18  id10344  We stayed here for - nights and were really ha...   
19  id10345  The service was fine, but the hotel itself fel...   
20  id10346  Having stayed at many Hilton properties, I exp...   
21  id10347  This is everything you could want from a hotel...   
22  id10348  I very much wanted to stay at a boutique hotel...   
23  id10349  I really liked this hotel. The staff were wond...   
24  id10350  My wife and spent - days there this month as a...   
25  id10351  We stayed at this hotel for two nights over th...   
26  id10352  Took a girls trip to LA and had no idea where ...   
27  id10353  Stayed with my girlfriend for a long weekend. ...   
28  id10354  I have stayed here numerous times and never be...   
29  id10355  The public areas are nice to look at. The staf...   

         Browser_Used Device_Used Is_Response  \
0                Edge      Mobile   not happy   
1   Internet Explorer      Mobile   not happy   
2             Mozilla      Tablet   not happy   
3    InternetExplorer     Desktop       happy   
4                Edge      Tablet   not happy   
5    InternetExplorer     Desktop       happy   
6             Firefox      Tablet   not happy   
7       Google Chrome      Mobile       happy   
8   Internet Explorer     Desktop       happy   
9              Chrome      Tablet   not happy   
10  Internet Explorer      Tablet       happy   
11                 IE      Mobile       happy   
12               Edge      Mobile   not happy   
13  Internet Explorer     Desktop       happy   
14             Chrome     Desktop       happy   
15            Mozilla     Desktop       happy   
16            Mozilla      Tablet   not happy   
17      Google Chrome     Desktop   not happy   
18      Google Chrome      Mobile       happy   
19            Mozilla     Desktop   not happy   
20            Firefox      Mobile   not happy   
21            Firefox     Desktop       happy   
22               Edge      Tablet   not happy   
23             Chrome     Desktop       happy   
24             Chrome     Desktop       happy   
25              Opera      Mobile   not happy   
26   InternetExplorer      Mobile       happy   
27    Mozilla Firefox     Desktop       happy   
28            Mozilla      Mobile       happy   
29            Firefox      Tablet   not happy   

                                          word_tokens  \
0   [room, kind, clean, had, very, strong, smell, ...   
1   [stayed, crown, plaza, april, april, staff, fr...   
2   [booked, hotel, hotwire, lowest, price, could,...   
3   [stayed, husband, sons, way, an, alaska, cruis...   
4   [girlfriends, stayed, celebrate, th, birthdays...   
5   [rooms, nice, clearly, had, been, updated, rec...   
6   [husband, i, stayed, this, hotel, 

In [14]:
print(df[df['compound_scores']>0.0000].shape[0])      #Positive  reviews
print(df[df['compound_scores']<0.0000].shape[0])      #Negative reviews
print(df[df['compound_scores']==0.0000].shape[0])     #Neutral reviews

5062
1296
32574


**Observation:**
- Out of the 38932 reviews, we can see that 5062 are positive (considering the compound factor), 1296 are negative and rest are neutral. 

In [20]:
#Check one positive score review description:
df.iloc[3]['Description']

'Stayed here with husband and sons on the way to an Alaska Cruise. We all loved the hotel, great experience. Ask for a room on the North tower, facing north west for the best views. We had a high floor, with a stunning view of the needle, the city, and even the cruise ships! We ordered room service for dinner so we could enjoy the perfect views. Room service dinners were delicious, too! You are in a perfect spot to walk everywhere, so enjoy the city. Almost forgot- Heavenly beds were heavenly, too!'

In [21]:
#Check one negative score review description:
df.iloc[9]['Description']

'Wonderful staff, great location, but it was definately the price was high for a standard hotel. The free breakfast was great, actually pretty good quailty for a free buffet. This hotel is in the heart of it all, walking distance to everything. - things, please Hyatt...the coffee was bad and the white ginger line of shampoo-conditioner-soap smells horrible!'

In [27]:
#Check one neutral score review description:
df.iloc[6]['Description']

"My husband and I have stayed in this hotel a few times. Though not the fanciest hotel, we love the fact that we can walk the - or so miles to Fenway. It is clean and the staff is very accomodating. My only complaint is that the fan in the bathroom was noisy and went on automatically when you turned the light on and we tried to keep the light off as much as possible. We've stayed in pricier hotels which charged for internet and breakfast and these are both included. Will stay there again."

# Hotel Review Sentiment Prediction: 
### TF-IDF model to vectorize the words and then applied Logistic regression to classify the reviews:
##### Hotel review 'Description' is the input avraible and the target variable is 'Is_Response'

In [113]:
df2 = df[['Description','Is_Response']]
df2.head()

Description Is_Response
0  The room was kind of clean but had a VERY stro...   not happy
1  I stayed at the Crown Plaza April -- - April -...   not happy
2  I booked this hotel through Hotwire at the low...   not happy
3  Stayed here with husband and sons on the way t...       happy
4  My girlfriends and I stayed here to celebrate ...   not happy

In [115]:
df2.shape[0]

1000

###### Text data pre-processing:

In [114]:
#clean the data: remove special charc, punctuation etc
import re
import string

#This function converts to lower-case, removes square bracket, removes numbers and punctuation
def text_clean_1(text):
    text = text.lower()
    text = re.sub('\[.*?\]', '', text)
    text = re.sub('[%s]' % re.escape(string.punctuation), '', text)
    text = re.sub('\w*\d\w*', '', text)
    return text

cleaned1 = lambda x: text_clean_1(x)


#store the cleaned data
df2['cleaned_description'] = pd.DataFrame(df2.Description.apply(cleaned1))
df2.head(10)

C:\Users\91842\anaconda3\lib\site-packages\ipykernel_launcher.py:17: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy


Description Is_Response  \
0  The room was kind of clean but had a VERY stro...   not happy   
1  I stayed at the Crown Plaza April -- - April -...   not happy   
2  I booked this hotel through Hotwire at the low...   not happy   
3  Stayed here with husband and sons on the way t...       happy   
4  My girlfriends and I stayed here to celebrate ...   not happy   
5  We had - rooms. One was very nice and clearly ...       happy   
6  My husband and I have stayed in this hotel a f...   not happy   
7  My wife & I stayed in this glorious city a whi...       happy   
8  My boyfriend and I stayed at the Fairmont on a...       happy   
9  Wonderful staff, great location, but it was de...   not happy   

                                 cleaned_description  
0  the room was kind of clean but had a very stro...  
1  i stayed at the crown plaza april   april   th...  
2  i booked this hotel through hotwire at the low...  
3  stayed here with husband and sons on the way t...  
4  my girlfriends and i stayed here to celebrate ...  
5  we had  rooms one was very nice and clearly ha...  
6  my husband and i have stayed in this hotel a f...  
7  my wife  i stayed in this glorious city a whil...  
8  my boyfriend and i stayed at the fairmont on a...  
9  wonderful staff great location but it was defi...

#### Train-test split:

In [108]:
from sklearn.model_selection import train_test_split

x = df2['cleaned_description']
y = df2['Is_Response']

x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.1, random_state = 225)

print('Independent Variable train :', len(x_train))
print('Independent Variable test:', len(x_test))
print('Dependent Variable train :', len(y_train))
print('Dependent Variable test  :', len(y_test))

Independent Variable train : 900
Independent Variable test: 100
Dependent Variable train : 900
Dependent Variable test  : 100


#### Model building:

In [109]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import LogisticRegression

tvec = TfidfVectorizer()
clf2 = LogisticRegression(solver = "lbfgs")

In [110]:
from sklearn.pipeline import Pipeline  
#pipeline will execute the steps 1 after another. At first, word vectors will be created using TF-IDF
#and then Logistic reg will be applied to classify the data based on the review descriptions

model = Pipeline([('vectorizer',tvec),('classifier',clf2)])     #First apply vectorizer and then classifier
model.fit(x_train, y_train)


from sklearn.metrics import confusion_matrix
predictions = model.predict(x_test)
confusion_matrix(predictions, y_test)

array([[65, 19],
       [ 2, 14]], dtype=int64)

##### Model Evaluation:

In [111]:
from sklearn.metrics import accuracy_score, precision_score, recall_score

print("Accuracy : ", accuracy_score(predictions, y_test))
print("Precision : ", precision_score(predictions, y_test, average = 'weighted'))
print("Recall : ", recall_score(predictions, y_test, average = 'weighted'))

Accuracy :  0.79
Precision :  0.8828041610131162
Recall :  0.79


##### Predict new reviews sentiment:

In [98]:
example1 = ["not clean at all,::Will not stay here. Disgusting."]
result = model.predict(example1)
print(result)

['not happy']


In [99]:
example2 = ["Loved the hotel, great experience"]
result2 = model.predict(example2)
print(result2)

['happy']
